### PR TITLE
Newfeature/layout comparator with treshold

### PR DESCRIPTION
**Pull request by plutasnyy**
I created error treshold

## Description

I added error treshold in pixels and percentages for screen comparator, also I had to create new status for frontend because we would like to know which test passed without changes and which test passed with changes. We changed css and html files for support new status with @m-suchorski . Finally I modified and added a little bit unit tests.

## Motivation and Context

For example, the time on the screenshot may be different, and the application may pass this test.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the [code style](https://github.com/Cognifide/aet/blob/master/CONTRIBUTING.md#coding-conventions) of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

---
I hereby agree to the terms of the AET Contributor License Agreement.

### CODE REVIEW

File: api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/ComparatorStepResult.java [hunk 1]
```diff
@@ -51,6 +51,7 @@ public boolean isRebaseable() {
     FAILED,
     WARNING,
     REBASED,
-    PROCESSING_ERROR
+    PROCESSING_ERROR,
+    CONDITIONAL,
```
Comment (tkaik) [inline api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/ComparatorStepResult.java]: I suggest changing it to `CONDITIONAL_PASS`

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java [hunk 2]
```diff
@@ -93,41 +106,95 @@ private ComparatorStepResult saveArtifacts(ImageComparisonResult imageComparison
...
+  private boolean isAcceptablePercentageChange(ImageComparisonResult mask) {
+    return mask.getPercentagePixelDifference() <= this.percentageThreshold.get();
+  }
+
+  private boolean isMaskWithoutDifference(ImageComparisonResult mask) {
+    return mask.getHeightDifference() == 0 && mask.getWidthDifference() == 0
+        && mask.getPixelDifferenceCount() == 0;
+  }
+
+  private void addPixelDifferenceDataToRestult(ComparatorStepResult result,
```
Comment (tkaik) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java]: Please fix the typo `Restult` -> `Result`

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java [hunk 3]
```diff
@@ -93,41 +106,95 @@ private ComparatorStepResult saveArtifacts(ImageComparisonResult imageComparison
...
     }
 
     return result;
   }
 
-  private ComparatorStepResult getPassedStepResult() {
-    ComparatorStepResult result = new ComparatorStepResult(null, ComparatorStepResult.Status.PASSED,
-        false);
+  boolean hasMaskThresholdWithAcceptableDifference(ImageComparisonResult mask) {
+    if (this.pixelThreshold.isPresent() && this.percentageThreshold.isPresent()) {
```
Comment (Skejven) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java]: `this` reference is unnecessary here, please remove it.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java [hunk 4]
```diff
@@ -93,41 +106,95 @@ private ComparatorStepResult saveArtifacts(ImageComparisonResult imageComparison
...
+          result = new ComparatorStepResult(maskArtifactId, Status.CONDITIONALLY_PASSED, true);
+        } else {
+          result = new ComparatorStepResult(maskArtifactId, Status.FAILED, true);
+        }
+        addPixelDifferenceDataToResult(result, imageComparisonResult);
+        addTimestampToResult(result);
       } catch (Exception e) {
         throw new ProcessingException(e.getMessage(), e);
       } finally {
+
```
Comment (Skejven) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java]: Remove this empty line.

File: report/src/main/webapp/assets/sass/_variables.scss [hunk 5]
```diff
@@ -60,6 +60,7 @@ $warning: #f0ad4e;
 $rebased: #0097fe;
 $failed: #bb5a5a;
 $passed: #6f9f00;
+$conditionallyPassed: #aacc2d;
```
Comment (Skejven) [inline report/src/main/webapp/assets/sass/_variables.scss]: I'm not sure about marking conditionally passed as a separate statistic and color. 
There should be proper filter to show conditionally passed, but let's not introduce additional color for now.

**Comment by tkaik** [comment]
As discussed with @Skejven, please add following changes on report level:
* number of PASS results should include conditional passed, e.g. change "0 (1)" to "1 (1)"
* use same color for conditional and regular pass (green)
* add different icon for Conditional Passed on both URL and Test level (e.g. https://fontawesome.com/icons/dot-circle?style=solid )
* add filter to "Show only conditionally passed" (existing "show passed urls" filter will also show conditionals)
CC @m-suchorski

File: core/jobs/src/test/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparatorTest.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  private LayoutComparator layoutComparator;
+
+  @Before
+  public void setUp() {
+    //given
+    this.layoutComparator = new LayoutComparator(this.comparatorProperties, this.artifactsDAO);
+  }
+
+  @Test
+  public void Comparator_hasMaskThresholdWithAcceptableDifference_withoutThreshold_expectFalse() {
```
Comment (tkaik) [inline core/jobs/src/test/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparatorTest.java]: `Comparator_` prefix is not needed in test names - please use [Tests naming convention](methodName_whenStateUnderTest_expectBehavior)

File: documentation/src/main/wiki/LayoutComparator.md [hunk 7]
```diff
@@ -10,7 +10,10 @@ Resource name: screen
 
 ##### Parameters
 
-No parameters
+| Parameter | Value | Description | Mandatory |
+| --------- | ----- | ----------- | --------- |
+| `pixelThreshold` | int (bigger than 0) | The value to set the error threshold in pixels | no |
+| `percentageThreshold` | double (0 to 100) | The value to set the error threshold in percentages | no |
```
Comment (tkaik) [inline documentation/src/main/wiki/LayoutComparator.md]: This documentation is missing some details - how the threshold mechanism works and what will happen on report if the difference will be below/equal/above threshold?

**Comment by tkaik** [comment]
Please update integration-tests with some example usages of new threshold parameters (e.g. in integration-tests/test-suite/partials/layout.xml file).

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java [hunk 8]
```diff
@@ -41,10 +43,18 @@
...
 
+  public static final String PERCENTAGE_THRESHOLD_PARAM = "percentageThreshold";
+
+  public static final String PIXEL_THRESHOLD_PARAM = "pixelThreshold";
+
   private final ComparatorProperties properties;
 
   private final ArtifactsDAO artifactsDAO;
 
+  private Integer pixelThreshold = null;
```
Comment (Skejven) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java]: You don't have to assign `null`. This is default value of java `Object`.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java [hunk 9]
```diff
@@ -116,18 +121,77 @@ private ComparatorStepResult saveArtifacts(ImageComparisonResult imageComparison
...
+    if (params.containsKey(PERCENTAGE_THRESHOLD_PARAM)) {
+      setPercentageThreshold(Double.valueOf(params.get(PERCENTAGE_THRESHOLD_PARAM)));
+      ParametersValidator
+          .checkRange(percentageThreshold.intValue(), 0, 100,
+              "Wrong percentage threshold value");
+    }
+    if (params.containsKey(PIXEL_THRESHOLD_PARAM)) {
+      setPixelThreshold(Integer.valueOf(params.get(PIXEL_THRESHOLD_PARAM)));
+      ParametersValidator.checkParameter(pixelThreshold >= 0,
+          "Pixel threshold should be greater than 0");
```
Comment (Skejven) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java]: Condition above says `greater or equal than 0` :)

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java [hunk 10]
```diff
@@ -116,18 +121,77 @@ private ComparatorStepResult saveArtifacts(ImageComparisonResult imageComparison
...
   }
 
   @Override
   public void setParameters(Map<String, String> params) throws ParametersException {
-    // no parameters needed
+    if (params.containsKey(PERCENTAGE_THRESHOLD_PARAM)) {
+      setPercentageThreshold(Double.valueOf(params.get(PERCENTAGE_THRESHOLD_PARAM)));
+      ParametersValidator
+          .checkRange(percentageThreshold.intValue(), 0, 100,
+              "Wrong percentage threshold value");
```
Comment (Skejven) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java]: Maybe that message would be more verbose:
> "Percentage threshold should be a decimal value between 0 and 100."

**Comment by tkaik** [comment]
Please also update documentation of Report webapp, e.g.:
- https://github.com/Cognifide/aet/wiki/SuiteReportFeatures#filtering-test-results - add "Conditionally passed" state
- https://github.com/Cognifide/aet/wiki/SuiteReportLayoutCase - describe what you can see when there is a conditional pass - e.g. images difference/mask is shown, but you can't accept is (test is passed)

File: integration-tests/test-suite/partials/layout.xml [hunk 11]
```diff
@@ -148,5 +148,117 @@
 			<url href="comparators/layout/success.jsp" />
 		</urls>
 	</test>
+
+	<test name="F-comparator-Layout-pixel-threshold">
```
Comment (tkaik) [inline integration-tests/test-suite/partials/layout.xml]: Please update expected number of failed/passed tests in bobcat tests (`integration-tests/sanity-functional`)

File: report/src/main/webapp/app/services/metadata.service.js [hunk 12]
```diff
@@ -287,6 +291,14 @@ define(['angularAMD', 'metadataCacheService', 'metadataEndpointService'],
           }
         }
 
+        function conditionallyPassed() {
+          decoratedObject.total++;
+          decoratedObject.conditionallyPassed++;
```
Comment (Skejven) [inline report/src/main/webapp/app/services/metadata.service.js]: Do we need to increase `passed` value here or not?

Comment (plutasnyy) [reply report/src/main/webapp/app/services/metadata.service.js]: We can increase `passed`. Now, it is evaluating there: https://github.com/Cognifide/aet/blob/f2dbc057f4278e8335583c8cfad43a43b0deb0ee/report/src/main/webapp/app/layout/sidepanel/sidepanel.view.html#L119
What do you think, which way will be better?
I am not 100% sure, why it works in toolbar:
https://github.com/Cognifide/aet/blob/f2dbc057f4278e8335583c8cfad43a43b0deb0ee/report/src/main/webapp/app/layout/toolbar/toolbarTop.view.html#L56
but as you can see on the screen shot, it works correctly:
https://github.com/Cognifide/aet/blob/f2dbc057f4278e8335583c8cfad43a43b0deb0ee/documentation/src/main/wiki/assets/suiteReport/layout-conditionally-passed.PNG
I think @m-suchorski know it better ;)

**Pull request merged**